app/models/filters/criteria.rb

```diff
@@ -0,0 +1,113 @@
+# Filters::Criteria provides a modular system for building and applying filter conditions to HMIS reporting queries.
+# It uses a factory pattern to instantiate criteria objects based on filter definitions, and provides utilities
+# for working with groups of related criteria.
+#
+# Each criteria class represents a specific type of filter (e.g., age range, project type, etc.) and follows
+# a consistent interface with `applies?` and `apply` methods.
+#
+# @example Creating and applying a single criterion
+#   filter = Filters::FilterBase.new(user_id: current_user.id, age_ranges: [:under_eighteen])
+#   criterion = Filters::Criteria.factory(:filter_for_age, input: filter)
+#   filtered_scope = criterion.apply(scope) if criterion.applies?
```

Comment:
Would it make sense to have `apply` just start with `return scope unless applies?`?
I can see the benefit of being explicit about checking applicability before calling apply, but since they are both instance methods, it seems like the behavior of `apply` when it doesn't apply, would just be to return the initial scope, unless it returns a `none` scope, but I _think_ that would be unexpected.

Reply:
Maybe. Right now calling apply() without checking applies? first has undefined behavior. If we want to be defensive we could add a guard against this if it seems like too much of a sharp edge. We could add that on the base class and require children to call super.

```ruby
  def applies?
    super && input.widget_id.present?
  end

  def apply(scope)
    # raises if applies? has not been called
    scope = super(scope)
    # ...filter scope
    scope.where(widget: input.widget_id)
  end
```